src/frontend/src/utils/countdown.ts

```diff
@@ -0,0 +1,102 @@
+import { render } from "lit-html";
+
+/**
+ * Countdown implementation which calls the supplied update function approximately every second until the endTimestamp is reached.
+ * When the endTimestamp is reached, the timeout callback supplied to the start method is called once.
+ * Stopping the countdown does not call the timeout callback.
+ */
+export class Countdown {
+  private interval = 1000;
+  private timeoutHandle: number | undefined;
+  private expected;
+  private readonly endTimestamp;
+
+  /**
+   * Constructs a new Countdown and starts it immediately.
+   * @param updateFunc Callback to be called approximately every second until endTimestamp is reached.
+   * @param endTimestamp timestamp (in nanoseconds) when the countdown should end.
+   * @param timeoutFunc Optional callback function to be called when the timeout is reached.
+   */
+  constructor(
+    private updateFunc: () => void,
+    endTimestamp: bigint,
+    private timeoutFunc?: () => Promise<void>
+  ) {
+    this.expected = Date.now() + this.interval;
+    this.endTimestamp = Number(endTimestamp / BigInt("1000000"));
+    this.timeoutHandle = window.setTimeout(() => this.step()); // execute the first step immediately
+  }
+
+  /**
+   * Stops the countdown. The timout callback is not called.
+   */
+  public stop(): void {
+    if (this.timeoutHandle !== undefined) {
+      window.clearTimeout(this.timeoutHandle);
+      this.timeoutHandle = undefined;
+    }
+  }
+
+  /**
+   * @return boolean whether the countdown is stopped (either because {@link stop} was called or because the timeout was reached).
+   */
+  public isStopped(): boolean {
+    return this.timeoutHandle !== undefined;
+  }
+
+  private async step() {
+    const now = Date.now();
+    if (now >= this.endTimestamp) {
+      if (this.timeoutFunc) {
+        await this.timeoutFunc();
+      }
+      return;
+    }
+    const drift = now - this.expected;
+    this.updateFunc();
+    this.expected += this.interval;
+    this.timeoutHandle = window.setTimeout(
+      () => this.step(),
+      Math.max(0, this.interval - drift)
+    );
+  }
+}
+
+/**
+ * Sets up a countdown ending at endTimestamp updating the timerElement every second.
+ * Note: this function will not yet start the countdown. To {@link Countdown.start} the countdown start() has to be called on the returned object.
+ * @param endTimestamp timestamp (in nanoseconds) when the countdown should end
+ * @param timerElement element to update with the remaining time in mm:ss format.
+ * @param timeoutFunc Optional callback function to be called when the timeout is reached.
+ * @return the initialized {@link Countdown}
+ */
+export const setupCountdown = (
+  endTimestamp: bigint,
+  timerElement: HTMLElement,
+  timeoutFunc?: () => Promise<void>
+): Countdown => {
+  return new Countdown(
+    () => render(formatRemainingTime(endTimestamp), timerElement),
```

Comment:
nit: this performs an actual DOM update every second, which might be a _bit_ slow, and is quite annoying when you have the dev tools open. Have you considered using a CSS variable and with the [CSS content attribute](https://www.w3schools.com/cssref/pr_gen_content.asp)? It might work though my CSS knowledge isn't great

Reply:
Well, mine isn't either. Let's file this under front-end improvements for later.